File: services/api-phoenix/lib/app/compliance.ex
defmodule App.Compliance do
  @moduledoc """
  Industry compliance profile, read at startup from compliance/profiles.json.
  One repo serves every industry; COMPLIANCE_PROFILE selects the active profile.

  profiles.json shape:
    %{
      "device" => "backend",
      "catalogVersion" => 1,
      "controlMeta" => %{...},
      "profiles" => %{
        "<profile>" => %{
          "name" => "...",
          "priority" => "...",
          "jurisdiction" => "...",
          "controls" => %{"<control>" => true | false | number | string}
        }
      }
    }

  Every profile carries the same control keys (uniform on/off catalog).
  """

  @default %{
    profile: "baseline",
    name: "Baseline (no industry lens)",
    jurisdiction: "Safe general defaults",
    controls: %{}
  }

  @doc "Load and cache the active profile. Call once at application start."
  def setup do
    profile = String.downcase(System.get_env("COMPLIANCE_PROFILE") || "baseline")
    :persistent_term.put(__MODULE__, load(profile))
  end

  @doc "The profile this process booted with."
  def active, do: :persistent_term.get(__MODULE__, @default)

  @doc """
  Read profiles.json and return the requested profile as a map of
  %{profile, name, jurisdiction, controls}. Fails loud on any error.
  """
  def load(profile) do
    path = Path.join([File.cwd!(), "compliance", "profiles.json"])

    with {:ok, text} <- File.read(path),
         {:ok, json} <- Jason.decode(text),
         %{} = profiles <- Map.get(json, "profiles"),
         %{} = entry <- Map.get(profiles, profile) do
      %{
        profile: profile,
        name: Map.get(entry, "name", profile),
        jurisdiction: Map.get(entry, "jurisdiction", ""),
        controls: Map.get(entry, "controls", %{})
      }
    else
      {:error, reason} ->
        # A profile catalog that cannot be read or parsed must fail loud.
        IO.puts(:stderr, "FATAL: compliance profiles.json not loadable: #{inspect(reason)}")
        System.halt(1)

      nil ->
        # A named profile with no matching entry must fail loud.
        IO.puts(:stderr, "FATAL: unknown COMPLIANCE_PROFILE: #{profile}")
        System.halt(1)
    end
  end
end

File: services/protocol-ws-node/src/compliance.ts
import { readFileSync } from 'node:fs'
import { join } from 'node:path'

// One repo serves every industry. The active profile is chosen at runtime by
// COMPLIANCE_PROFILE — no code change, no rebuild. The profiles come from the
// platform's canonical catalog (compliance/profiles.json), generated to JSON so
// there is no per-language YAML parsing.

export interface ComplianceView {
  profile: string
  name: string
  jurisdiction: string
  // Uniform control set for this device type. Same keys for every profile.
  controls: Record<string, string | number | boolean>
}

interface ProfilesFile {
  device: string
  catalogVersion: number
  controlMeta: Record<string, { label: string; type: string }>
  profiles: Record<string, {
    name: string; priority: string; jurisdiction: string
    controls: Record<string, string | number | boolean>
  }>
}

let cache: ProfilesFile | null = null
function loadProfiles(): ProfilesFile | null {
  if (cache) return cache
  try {
    cache = JSON.parse(
      readFileSync(join(process.cwd(), 'compliance', 'profiles.json'), 'utf8'),
    ) as ProfilesFile
    return cache
  } catch {
    return null
  }
}

export function activeCompliance(): ComplianceView {
  const profile = (process.env.COMPLIANCE_PROFILE ?? 'baseline').toLowerCase()
  const data = loadProfiles()
  const p = data?.profiles[profile] ?? data?.profiles.baseline
  if (!p) return { profile, name: 'unknown', jurisdiction: '', controls: {} }
  return { profile, name: p.name, jurisdiction: p.jurisdiction, controls: p.controls }
}
